src/huggingface_hub/serialization/_torch.py

```diff
@@ -349,6 +354,336 @@ def split_torch_state_dict_into_shards(
+    model: "torch.nn.Module",
+    checkpoint_path: Union[str, os.PathLike],
+    *,
+    strict: bool = False,
+    safe: bool = True,
+    weights_only: bool = False,
+    map_location: Optional[Union[str, "torch.device"]] = None,
+) -> NamedTuple:
+    """
+    Load a checkpoint into a model, handling both sharded and non-sharded checkpoints.
+
+    Args:
+        model (`torch.nn.Module`):
+            The model in which to load the checkpoint.
+        checkpoint_path (`str` or `os.PathLike`):
+            Path to either the checkpoint file or directory containing the checkpoint(s).
+        strict (`bool`, *optional*, defaults to `False`):
+            Whether to strictly enforce that the keys in the model state dict match the keys in the checkpoint.
+        safe (`bool`, *optional*, defaults to `True`):
+            If both safetensors and PyTorch save files are present in checkpoint and `safe` is True, the
+            safetensors files will be loaded. Otherwise, PyTorch files are always loaded when possible.
+        weights_only (`bool`, *optional*, defaults to `False`):
+            If True, only loads the model weights without optimizer states and other metadata.
+            Only supported in PyTorch >= 1.13.
+        map_location (`str` or `torch.device`, *optional*):
+            A `torch.device` object, string or a dict specifying how to remap storage locations. It
+            indicates the location where all tensors should be loaded.
+
+    Returns:
+        `NamedTuple`: A named tuple with `missing_keys` and `unexpected_keys` fields.
+            - `missing_keys` is a list of str containing the missing keys, i.e. keys that are in the model but not in the checkpoint.
+            - `unexpected_keys` is a list of str containing the unexpected keys, i.e. keys that are in the checkpoint but not in the model.
+
+    Raises:
+        ValueError: If the checkpoint path is invalid or if the checkpoint format cannot be determined.
+        FileNotFoundError: If the checkpoint file or directory does not exist.
+        ImportError: If required dependencies are not installed.
+
+    Example:
+    ```python
+    >>> from huggingface_hub import load_torch_model
+    >>> model = ... # A PyTorch model
+    >>> load_torch_model(model, "path/to/checkpoint")
+    ```
+    """
+    checkpoint_path = Path(checkpoint_path)
+
+    if checkpoint_path.is_file():
+        state_dict = load_state_dict_from_file(
+            checkpoint_file=checkpoint_path,
+            map_location=map_location,
+            weights_only=weights_only,
+        )
+        return model.load_state_dict(state_dict, strict=strict)
+
+    elif checkpoint_path.is_dir():
+        # Check for index file which indicates sharded checkpoint
+        filename_pattern = (
+            constants.SAFETENSORS_WEIGHTS_FILE_PATTERN if safe else constants.PYTORCH_WEIGHTS_FILE_PATTERN
+        )
+        index_file = checkpoint_path / (filename_pattern.format(suffix="") + ".index.json")
```

Comment:
Here if `safe=True`, it's clear that we want to load from `safetensors`. But if `safe=False`, we could still try to load from `safetensors` if it's installed and the weights are available in that format. So maybe we could have a logic like (pseudo-code):

```py
index_file = checkpoint_path / (constants.SAFETENSORS_WEIGHTS_FILE_PATTERN.format(suffix="") + ".index.json")
if not index_file.is_file() and safe:
    index_file := checkpoint_path / (constants.PYTORCH_WEIGHTS_FILE_PATTERN.format(suffix="") + ".index.json")
if index_file.is_file():
    ... # load

... # else, same as now: check if single file exists or raise error
```

Reply:
The logic to prioritize loading safetensors makes sense since the [docstring](https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/serialization/_torch.py#L97) for `safe_serialization` states that "Saving as pickle is deprecated and will be removed in a future version."
this is addressed in [0cce776](https://github.com/huggingface/huggingface_hub/pull/2687/commits/0cce776851f707a8ac9a7fc0fbfba4353882384b), I'd appreciate your opinion on this specific part of the code!